src/compute/src/render/reduce.rs

```diff
@@ -837,29 +839,30 @@ where
         (key, values)
     });
 
-    // We can place our rows directly into the diff field, and only keep the
-    // relevant one corresponding to evaluating our aggregate, instead of having
-    // to do a hierarchical reduction.
-    use timely::dataflow::operators::Map;
-
+    let debug_name = debug_name.to_string();
+    let (partial, errs) = collection.ensure_monotonic(move |data, diff| {
+        warn!("[customer-data] ReduceMonotonic expected monotonic input but received {data:?} with diff {diff:?} in dataflow {debug_name}");
+        error!("Non-monotonic input to ReduceMonotonic");
+        let m = "tried to build a monotonic reduction on non-monotonic input".to_string();
+        (DataflowError::from(EvalError::Internal(m)), 1)
+    });
```

Comment:
You might be able to select from one of the tables that have a long retention period. [`mz_cluster_replica_utilization`](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_cluster_replica_utilization) has a retention of 30 days, and reading from it with an as-of in the past would currently guarantee that the persist source replays a lot of history. That said, there might be easier ways to emulate this behavior, similar to the tricks @vmarcos was using to force invalid data into reductions for testing purposes.

Reply:
Oh yeah, I had considered `repeat_row` as a strategy for producing bad input, but I hadn't thought about using it to create something that's good, but only once consolidated.  That's worth a shot.

Reply:
Unfortunately, this is indeed a bit tricky to test currently. The problem is that `repeat_row` is marked to be monotonicity-breaking: https://github.com/MaterializeInc/materialize/blob/bae5ea5cfe28408bad1fe64e99a17b16803f099d/src/expr/src/relation/func.rs#L2510
It would be possible to do some manual testing by turning the above to `true`; not sure how I would feel about flipping it to `true` more definitely. It might be OK if we deem `repeat_row` a testing-only function intended to create incorrect behavior? Strictly, it is not monotonicity-preserving, though.
Another option is to add automated tests later when we work on #17358?

Reply:
> Unfortunately, this is indeed a bit tricky to test currently. The problem is that `repeat_row` is marked to be monotonicity-breaking:

Alas, that explains the failure of my attempts to trigger this.  

> Another option is to add automated tests later when we work on #17358?

Agreed.  I think I will merge this for now, but continue thinking about the best way to approach testing all these error paths in general.